### DISCUSSION TITLE
Dynamic workflow with timers

### DISCUSSION THREAD

**Discussion by elpool** [elsa-workflows/elsa-core#2847] [Q&A]
I am trying to build a workflow blueprint dynamically, as suggested in this other discussion here: #2835 
I was able to follow the example but for some reason I can't get _timers_ to work.

The Timer activity works correctly when the workflow is defined using IWorkflowBuilder, but when defined as below it never actually resumes

```
var workflowDefinition = new WorkflowDefinition()
{
	Id = "1",
	DefinitionId = "SampleWorkflow",
	Version = 1,
	IsPublished = true,
	IsLatest = true,
	PersistenceBehavior = WorkflowPersistenceBehavior.Suspended,
};

workflowDefinition.Variables.Set("Test", "TestValue");

var step1 = new ActivityDefinition()
{
	ActivityId = Guid.NewGuid().ToString("N"),
	Type = "SayHelloWorld",
	Name = "Step1",
	Properties = new List<ActivityDefinitionProperty>
	{
		ActivityDefinitionProperty.Liquid("Message", "here is a {{ Variables.Test }}"),
	}
};
workflowDefinition.Activities.Add(step1);

var step2 = new ActivityDefinition()
{
	ActivityId = Guid.NewGuid().ToString("N"),
	Type = "Timer",
	Name = "Step2",
	Properties = new List<ActivityDefinitionProperty>
	{
		ActivityDefinitionProperty.Literal("Timeout", "0:00:00:10"),
	}
};
workflowDefinition.Activities.Add(step2);

var step3 = new ActivityDefinition()
{
	ActivityId = Guid.NewGuid().ToString("N"),
	Type = "SayHelloWorld",
	Name = "Step3",
	Properties = new List<ActivityDefinitionProperty>
	{
		ActivityDefinitionProperty.Liquid("Message", "This came from step 1: {{ Activities.Step1.Output }}"),
	}
};
workflowDefinition.Activities.Add(step3);

workflowDefinition.Connections.Add(new ConnectionDefinition(step1.ActivityId, step2.ActivityId, "Done"));
workflowDefinition.Connections.Add(new ConnectionDefinition(step2.ActivityId, step3.ActivityId, "Done"));
```

Any idea what is causing this?

**Comment by sfmskywalker** [comment]
A couple of things to look for:

1. Since this workflow does not start with the Timer activity (but with the SayHelloWorld activity), this workflow has to be invoked explicitly (i.e. it won't automatically execute since the Timer is not acting as a workflow trigger here).
2. Timers only work if the application is run from a .NET host. I.e. it won't work if you have a console application from which you execute the workflow.

**Reply by elpool** [reply]
ok thanks for the help, I think I am doing both of those things.
Here's how I'm invoking the workflow:
```
var host = new HostBuilder()
	.ConfigureServices(ConfigureServices)
	.UseConsoleLifetime()
	.Build();

using (host)
{
	await host.StartAsync();
 
        // If I start a workflow defined in a IWorkflow class that includes a timer step, it works: 
	//var workflowRunner = host.Services.GetRequiredService<IBuildsAndStartsWorkflow>();
	//await workflowRunner.BuildAndStartWorkflowAsync<HelloWorld>();

        // But if I start a workflow from an IWorkflowBlueprint it runs the first (non-blocking) activity but the timer never fires
	var workflowRunner = host.Services.GetRequiredService<IStartsWorkflow>();
	IWorkflowBlueprint myWorkflow = await BuildWorkflow(host.Services);
	await workflowRunner.StartWorkflowAsync(myWorkflow);

	await host.WaitForShutdownAsync();
}
```

thanks again for your help

**Reply by sfmskywalker** [reply]
I'm not sure why it doesn't work when sing the workflow definition you dynamically built, but one other reason might be that it does not get registered with the workflow registry (which is mandatory for long-running processes if they are to be resumed by the time management service). But, the same would go for `IWorkflow` implementations, so this explanation wouldn't make much sense.

Perhaps you can send me a (stripped down) version of your project so I can take a quick look?
